4/m_censor.cpp

```diff
@@ -42,22 +36,166 @@ class ModuleCensor : public Module
+
+			if (std::isalpha(static_cast<unsigned char>(c)))
+			{
+				ScriptType current = std::islower(static_cast<unsigned char>(c)) || std::isupper(static_cast<unsigned char>(c)) ? SCRIPT_LATIN : SCRIPT_NONLATIN;
+				if (detected == SCRIPT_UNKNOWN)
+				{
+					detected = current;
+				}
+				else if (detected != current)
+				{
+					return true; // Mixed scripts detected
+				}
+			}
+		}
+
+		return false;
+	}
+
+	// Helper function to convert UTF-8 string to UTF-32
+	std::u32string to_utf32(const std::string& utf8)
+	{
+		std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> convert;
+		return convert.from_bytes(utf8);
+	}
+
+	// Helper function to convert UTF-32 character to UTF-8
+	std::string to_utf8(char32_t utf32_char)
+	{
+		std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> convert;
+		return convert.to_bytes(utf32_char);
+	}
+
+	bool IsEmojiOnly(const std::string& text)
+	{
+		UErrorCode status = U_ZERO_ERROR;
+		icu::UnicodeString ustr(text.c_str(), "UTF-8");
+		std::unique_ptr<icu::RegexMatcher> emoji_matcher(emoji_pattern->matcher(ustr, status));
+		if (U_FAILURE(status))
+		{
+			ServerInstance->Logs.Normal(MODNAME, "Failed to create regex matcher for emojis: %s", u_errorName(status));
+			return false;
+		}
+
+		// Check if the entire text is matched by the emoji pattern
+		return emoji_matcher->matches(status);
+	}
+
+	bool IsKiwiIRCOnly(const std::string& text)
+	{
+		UErrorCode status = U_ZERO_ERROR;
+		icu::UnicodeString ustr(text.c_str(), "UTF-8");
+		std::unique_ptr<icu::RegexMatcher> kiwiirc_matcher(kiwiirc_pattern->matcher(ustr, status));
+		if (U_FAILURE(status))
+		{
+			ServerInstance->Logs.Normal(MODNAME, "Failed to create regex matcher for KiwiIRC: %s", u_errorName(status));
+			return false;
+		}
+
+		// Check if the entire text is matched by the KiwiIRC pattern
+		return kiwiirc_matcher->matches(status);
+	}
+
+	bool IsAllowed(const std::string& text)
+	{
+		// Allow ASCII characters and common symbols by default
+		if (std::all_of(text.begin(), text.end(), [](unsigned char c) { return c >= 32 && c <= 126; }))
+		{
+			return true;
+		}
+
+		UErrorCode status = U_ZERO_ERROR;
+		icu::UnicodeString ustr(text.c_str(), "UTF-8");
+		std::unique_ptr<icu::RegexMatcher> whitelist_matcher(whitelist_pattern->matcher(ustr, status));
+		if (U_FAILURE(status))
+		{
+			ServerInstance->Logs.Normal(MODNAME, "Failed to create regex matcher for whitelist: %s", u_errorName(status));
+			return false;
+		}
+
+		return whitelist_matcher->matches(status) || IsEmojiOnly(text) || IsKiwiIRCOnly(text);
+	}
 
  public:
 	ModuleCensor()
-		: Module(VF_NONE, "Allows the server administrator to define inappropriate phrases that are not allowed to be used in private or channel messages.")
+		: Module(VF_NONE, "Allows the server administrator to define inappropriate phrases that are not allowed to be used in private or channel messages and blocks messages with mixed UTF-8 scripts, only allowing certain Unicode smileys.")
```

Comment:
This should probably just be a new module rather than tacking it on to an existing one (which needs to keep compatibility with the v3 version).